cors/mod.ts

```diff
@@ -0,0 +1,207 @@
+  optionsSuccessStatus?: number;
+  preflight?: boolean;
+};
+
+const isOriginAllowed = (reqOrigin: string, origin: OriginOption): boolean => {
+  if (Array.isArray(origin)) {
+    for (let i = 0; i < origin.length; ++i) {
+      if (isOriginAllowed(reqOrigin, origin[i])) {
+        return true;
+      }
+    }
+
+    return false;
+  } else if (typeof origin === "string") {
+    return reqOrigin === origin;
+  } else if (origin instanceof RegExp) {
+    return origin.test(reqOrigin);
+  } else {
+    return !!origin;
+  }
+};
+
+export const CorsMiddleware = ({
+  origin = "*",
+  credentials = false,
+  exposeHeaders,
+  allowMethods = "GET,HEAD,PUT,PATCH,POST,DELETE",
+  allowHeaders,
+  maxAge,
+  optionsSuccessStatus = 204,
+  preflight = true,
+}: CorsMiddlewareConfig = {}) =>
+  async (
+    request: Drash.Http.Request,
+    response: Drash.Http.Response,
+  ): Promise<void> => {
+    let options: any = {
+      origin,
+      credentials,
+      optionsSuccessStatus,
+      preflight,
+    };
+
+    if (!options.origin) {
+      return;
+    }
+
+    if (Array.isArray(exposeHeaders)) {
+      options.exposeHeaders = exposeHeaders.join(",");
+    } else {
+      options.exposeHeaders = exposeHeaders;
+    }
+
+    if (Array.isArray(allowMethods)) {
+      options.allowMethods = allowMethods.join(",");
+    } else {
+      options.allowMethods = allowMethods;
+    }
+
+    if (Array.isArray(allowHeaders)) {
+      options.allowHeaders = allowHeaders.join(",");
+    } else {
+      options.allowHeaders = allowHeaders;
+    }
+
+    if (maxAge) {
+      options.maxAge = String(maxAge);
+    }
+
+    // If the Origin header is not present terminate this set of steps.
+    // The request is outside the scope of this specification.
+    const requestOrigin = request.headers.get("Origin");
+
+    // Always set Vary header
+    // https://github.com/rs/cors/issues/10
+    vary(
+      (header: string) => response.headers.get(header) || "",
+      (header: string, value: string) => {
+        response.headers.set(header, value);
+      },
+      "origin",
+    );
+    if (!requestOrigin) {
+      return;
+    }
+
+    let o;
+
+    if (!options.origin || options.origin === "*") {
+      // allow any origin
+      o = "*";
+    } else if (typeof options.origin === "string") {
+      // fixed origin
+      o = options.origin;
+    } else if (typeof options.origin === "function") {
+      o = options.origin(request, response);
+
+      if (o instanceof Promise) {
+        o = await o;
+      }
+
+      if (!o) {
+        // Safari (and potentially other browsers) need content-length 0,
+        //   for 204 or they just hang waiting for a body
+        response.status_code = options.optionsSuccessStatus as number;
+        response.headers.set("Content-Length", "0");
+
+        return;
+      }
+    } else {
+      // reflect origin
+      o = isOriginAllowed(requestOrigin, options.origin) ? requestOrigin
+      : false;
+    }
+
+    let c;
```

Comment:
Can this variable be named so it's descriptive of it's purpose?

Reply:
https://github.com/drashland/deno-drash-middleware/pull/1/commits/1627c064bf674eeeccbe726cd79bd685da78dd8b